File: ios/Wrappers/ConversationWrapper.swift
import Foundation
import XMTP

// Wrapper around XMTP.Conversation to allow passing these objects back into react native.
struct ConversationWrapper {
	static func encodeToObj(_ conversation: XMTP.Conversation, client: XMTP.Client, conversationParams: ConversationParamsWrapper = ConversationParamsWrapper()) async throws -> [String: Any] {
		switch conversation {
		case .group(let group):
			return try await GroupWrapper.encodeToObj(group, client: client, conversationParams: conversationParams)
		case .dm(let dm):
			return try await DmWrapper.encodeToObj(dm, client: client, conversationParams: conversationParams)
		}
	}
    static func encode(_ conversation: XMTP.Conversation, client: XMTP.Client, conversationParams: ConversationParamsWrapper = ConversationParamsWrapper()) async throws -> String {
        let obj = try await encodeToObj(conversation, client: client, conversationParams: conversationParams)
		let data = try JSONSerialization.data(withJSONObject: obj)
		guard let result = String(data: data, encoding: .utf8) else {
			throw WrapperError.encodeError("could not encode conversation")
		}
		return result
	}
}

File: android/src/main/java/expo/modules/xmtpreactnativesdk/wrappers/ConversationWrapper.kt
package expo.modules.xmtpreactnativesdk.wrappers

import com.google.gson.GsonBuilder
import org.xmtp.android.library.Client
import org.xmtp.android.library.Conversation

class ConversationWrapper {
    companion object {
        suspend fun encodeToObj(
            client: Client,
            conversation: Conversation,
            conversationParams: ConversationParamsWrapper = ConversationParamsWrapper(),
        ): Map<String, Any?> =
            when (conversation.type) {
                Conversation.Type.GROUP -> {
                    val group = (conversation as Conversation.Group).group
                    GroupWrapper.encodeToObj(client, group, conversationParams)
                }

                Conversation.Type.DM -> {
                    val dm = (conversation as Conversation.Dm).dm
                    DmWrapper.encodeToObj(client, dm, conversationParams)
                }
            }

        suspend fun encode(
            client: Client,
            conversation: Conversation,
            conversationParams: ConversationParamsWrapper = ConversationParamsWrapper(),
        ): String {
            val gson = GsonBuilder().create()
            val obj =
                encodeToObj(client, conversation, conversationParams)
            return gson.toJson(obj)
        }
    }
}
